flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataTypeQueryable.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * Indicates that a {@link DataType} can be retrieved from a class that implements this interfaces.
+ */
+@PublicEvolving
+public interface DataTypeQueryable {
```

Comment:
Is this interface really necessary? Especially with `@PublicEvolving` annotation? How are users supposed to use it? If I understand it correctly you need it for internal operations. Moreover you need it because the `WrapperTypeInfo` is in `blink-runtime`, right?

Can't we move the `WrapperTypeInfo` to the `table-common` instead?  The class itself has no runtime dependencies. Only the factory methods need some runtime classes.

Reply:
I thought about this as well and decided for keeping `WrapperTypeInfo` in the runtime package. The reason for this is to keep the serializers close to this utility class in the same package. `WrapperTypeInfo` is not needed in `table-common` because it is internal. Actually the constructor of `WrappenTypeInfo` should even be private but this is not possible due to other failing tests so far. We can mark `DataTypeQueryable` internal if that helps, but the past around `TypeInformation` has shown that you need such a simple interface over and over again. It will be more useful in the future once we introduce a "converter type serializer" where the return data type is not of internal data structure.

Reply:
I prefer to keep `WrapperTypeInfo` in `blink-runtime`, especially we have a public construct which will cause a lot of problems if the `logicalType` and `typeSerializer` are not mapped.

Reply:
Is it possible to make the construct of `WrapperTypeInfo` to be package visible? We can move the utility `TypeInfoDataTypeConverter#createWrapperTypeInfo` into `WrapperTypeInfo`. And the remaining constructor invoking are the tests under the same package.